src/ILLink.Shared/XmlAttributeParser.cs

```diff
@@ -0,0 +1,317 @@
+					Parameters.Add (new ParameterNode (parameterNav));
+				}
+				foreach (XPathNavigator returnNav in nav.SelectChildren (ReturnElementName, "")) {
+					foreach (XPathNavigator attributeNav in returnNav.SelectChildren (AttributeElementName, "")) {
+						ReturnAttributes.Add (new AttributeNode (attributeNav));
+					}
+				}
+			}
+		}
+
+		public abstract record TypeBaseNode : FeatureSwitchedNode
+		{
+			public List<TypeMemberNode> Events;
+			public List<TypeMemberNode> Fields;
+			public List<TypeMemberNode> Properties;
+			public List<MethodNode> Methods;
+			public List<NestedTypeNode> Types;
+			public TypeBaseNode (XPathNavigator nav) : base (nav)
+			{
+				Methods = new List<MethodNode> ();
+				Properties = new List<TypeMemberNode> ();
+				Fields = new List<TypeMemberNode> ();
+				Events = new List<TypeMemberNode> ();
+				foreach (XPathNavigator methodNav in nav.SelectChildren (MethodElementName, "")) {
+					Methods.Add (new MethodNode (methodNav));
+				}
+				foreach (XPathNavigator propertyNav in nav.SelectChildren (PropertyElementName, "")) {
+					Properties.Add (new TypeMemberNode (propertyNav));
+				}
+				foreach (XPathNavigator eventNav in nav.SelectChildren (EventElementName, "")) {
+					Events.Add (new TypeMemberNode (eventNav));
+				}
+				foreach (XPathNavigator fieldNav in nav.SelectChildren (FieldElementName, "")) {
+					Fields.Add (new TypeMemberNode (fieldNav));
+				}
+				Types = new List<NestedTypeNode> ();
+				foreach (XPathNavigator nestedTypeNav in nav.SelectChildren (TypeElementName, "")) {
+					Types.Add (new NestedTypeNode (nestedTypeNav));
+				}
+			}
+		}
+		public partial record NestedTypeNode : TypeBaseNode
+		{
+			public string Name;
+			public NestedTypeNode (XPathNavigator nav) : base (nav)
+			{
+				Name = GetName (nav);
+			}
+		}
+
+		public partial record TypeNode : TypeBaseNode, ITopLevelNode
+		{
+			public string FullName;
+			public TypeNode (XPathNavigator nav) : base (nav)
+			{
+				FullName = GetFullName (nav);
+			}
+		}
+
+		public record AssemblyNode : FeatureSwitchedNode, ITopLevelNode
+		{
+			public List<TypeNode> Types;
+			public string FullName;
+			public AssemblyNode (XPathNavigator nav) : base (nav)
+			{
+				FullName = GetFullName (nav);
+				Types = new List<TypeNode> ();
+				foreach (XPathNavigator typeNav in nav.SelectChildren (TypeElementName, "")) {
+					Types.Add (new TypeNode (typeNav));
+				}
+			}
+		}
+
+		public interface ITopLevelNode
+		{
+		}
+
+		public abstract record NodeBase : XmlProcessorBase
+		{
+			public IXmlLineInfo? LineInfo;
+			protected NodeBase (XPathNavigator nav)
+			{
+				LineInfo = (nav is IXmlLineInfo lineInfo) ? lineInfo : null;
+			}
+
+			public static List<ITopLevelNode> ProcessRootNodes (XPathNavigator nav)
+			{
+				if (!nav.MoveToChild (LinkerElementName, XmlNamespace)) {
+					throw new ArgumentException ($"XML does not have <{LinkerElementName}> base tag");
```

Comment:
Generally we shouldn't throw exceptions on malformed XML. Is the idea here that the schema validation will cover it already, and we can never get here for an XML which doesn't have `<linker>` at the root? If that's the case it seems fine.

Reply:
I removed exceptions from xml parsing, so I'll resolve this conversation.